torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -0,0 +1,57 @@
+from typing import Dict, List
+
+import torch
+import torch.nn as nn
+
+from torch.distributed.fsdp._common_utils import _named_parameters_with_duplicates
+
+from ._fsdp_common import FSDPMeshInfo
+from ._fsdp_param import FSDPParam, ParamModuleInfo
+
+
+class FSDPParamGroup:
+    """This class represents a parameter group to communicate together."""
+
+    def __init__(
+        self,
+        params: List[nn.Parameter],
+        module: nn.Module,
+        mesh_info: FSDPMeshInfo,
+        device: torch.device,
+    ):
+        self.module = module  # permit ref cycle because 1:1 lifetime
+        param_module_infos = _get_param_module_infos(params, module)
+        self.fsdp_params = [
+            FSDPParam(param, module_info, mesh_info, device)
+            for param, module_info in zip(params, param_module_infos)
+        ]
+        self.mesh_info = mesh_info
+        self.device = device
+
+
+def _get_param_module_infos(
+    params: List[nn.Parameter], module: nn.Module
+) -> List[ParamModuleInfo]:
+    """
+    Shared parameter:
+        lin1.weight = lin2.weight
+    Shared module:
+        mlp.lin1 = mlp.lin2
+    We do not remove duplicates when traversing both modules and parameters to
+    find shared modules' parameters and shared parameters within a module.
+    """
+    params_set = set(params)
+    param_to_module_info: Dict[nn.Parameter, ParamModuleInfo] = {}
+    for _, submodule in module.named_modules(remove_duplicate=False):
+        for param_name, param in _named_parameters_with_duplicates(
+            submodule, recurse=False
+        ):
+            if param in params_set:
+                if param not in param_to_module_info:
+                    param_to_module_info[param] = ParamModuleInfo(submodule, param_name)
+                else:
+                    param_to_module_info[param].shared_modules.append(submodule)
+                    param_to_module_info[param].shared_param_names.append(param_name)
```

Comment:
I think a `Dict[str, nn.Module]` may not work for this usage. We want the unprefixed parameter name (e.g. `weight`, not `lin.weight`) so that we can `setattr(module, param_name, tensor)` with the unprefixed `param_name`.

However, we could have a case like: `lin1.weight = lin2.weight = lin3.weight`, in which case all of the `param_name`s are `"weight"` and `"weight"` will no longer be a unique key. Alternatively, we could `.split('.')[-1]` each time we use the parameter name, but the current 2-list approach might be more direct.